test/test_fake_quant.py

```diff
@@ -5,143 +5,103 @@
 import torch.jit
 import numpy as np
 import unittest
+from hypothesis import assume, given
+from hypothesis import strategies as st
 from common_utils import run_tests
-import torch.nn._intrinsic as _intrinsic
+from torch.quantization import FakeQuantize
+from torch.quantization import default_qconfig
 
 
 # Reference method for quantizing a tensor.
 def _fake_quantize_per_tensor_affine_reference(X, scale, zero_point, quant_min, quant_max):
-    res = (np.clip(np.round(X / scale) + zero_point, quant_min, quant_max) - zero_point) * scale
+    res = (torch.clamp(torch.round(X / scale) + zero_point, quant_min, quant_max) - zero_point) * scale
     res = res.reshape(X.shape)
     return res
 
 
 # Reference method for the gradient of the quantizer.
-def _fake_quantize_per_tensor_affine_grad_reference(X, dY, scale, zero_point, quant_min, quant_max):
-    Xq = np.round(X / scale) + zero_point
-    mask = np.logical_and(Xq >= quant_min, Xq <= quant_max)
-    res = dY[mask].reshape(dY.shape)
+def _fake_quantize_per_tensor_affine_grad_reference(dY, X, scale, zero_point, quant_min, quant_max):
+    Xq = torch.round(X / scale) + zero_point
+    mask = (Xq >= quant_min) & (Xq <= quant_max)
+    res = torch.zeros_like(dY)
+    res[mask] = dY[mask]
     return res
 
 NP_RANDOM_SEED = 19
-
+tolerance = 1e-6
 
 class TestFakeQuantizePerTensorAffine(unittest.TestCase):
+    def to_tensor(self, X, device):
```

Comment:
General comment for all tests: We need to test saturation within the quantization for both forward and backward. This can be done by setting scale to a small value: For the example below, data is uniform in [0,125], so a scale of less than 125/255 would work.  Can we add randomized testing here to make sure that we get sufficient coverage?

Reply:
switched to using hypothesis_utils